codex/stores/maintenance.nim

```diff
@@ -10,90 +10,318 @@
+          raise newException(CatchableError, "Progress should be increasing only, treeCid " & $treeCid)
+
+        if currDatasetMd.leavesCount <= datasetMd.checkpoint.progress:
+          DatasetMetadata.none
+        else:
+          datasetMd.some
+      else:
+        raise newException(CatchableError, "Metadata for dataset with treeCid " & $treeCid & " not found")
   )
 
-  without iter =? expirations, err:
-    trace "Unable to obtain blockExpirations iterator from repoStore"
-    return
+proc deleteBatch(self: DatasetMaintainer, treeCid: Cid, datasetMd: DatasetMetadata): Future[?!void] {.async.} =
+  var datasetMd = datasetMd
+
+  datasetMd.checkpoint.timestamp = await self.clock.now
+
+  if err =? (await self.recordCheckpoint(treeCid, datasetMd)).errorOption:
+    return failure(err)
+
+  logScope:
+    treeCid = treeCid
+    manifestsCids = datasetMd.manifestsCids
+    startingIndex = datasetMd.checkpoint.progress
+
+  trace "Deleting a batch of blocks", size = self.batchSize
+
+  var index = datasetMd.checkpoint.progress
+  while (index < datasetMd.checkpoint.progress + self.batchSize) and
+          (index < datasetMd.leavesCount):
+    if err =? (await self.blockStore.delBlock(treeCid, index)).errorOption:
+      error "Error deleting a block", msg = err.msg, index
+
+    index.inc
+
+    await sleepAsync(50.millis) # cooperative scheduling
+
+  if index >= datasetMd.leavesCount:
+    trace "All blocks deleted from a dataset", leavesCount = datasetMd.leavesCount
+
+    for manifestCid in datasetMd.manifestsCids:
+      if err =? (await self.blockStore.delBlock(manifestCid)).errorOption:
+        error "Error deleting manifest", cid = manifestCid
+
+    if err =? (await self.blockStore.delBlock(treeCid)).errorOption:
+      error "Error deleting block", cid = treeCid
+
+    if err =? (await self.recordCheckpoint(treeCid, datasetMd)).errorOption:
+      return failure(err)
+
+    return success()
+  else:
+    datasetMd.checkpoint.progress = index
+    return await self.deleteBatch(treeCid, datasetMd)
+
+proc superviseDatasetDeletion(self: DatasetMaintainer, treeCid: Cid, datasetMd: DatasetMetadata): Future[void] {.async.} =
+  logScope:
+    treeCid = treeCid
+    manifestsCids = datasetMd.manifestsCids
+    expiry = datasetMd.expiry
+    leavesCount = datasetMd.leavesCount
+
+  try:
+    if datasetMd.checkpoint.progress == 0 and datasetMd.checkpoint.timestamp == 0:
+      info "Initiating deletion of a dataset"
+    else:
+      info "Retrying deletion of a dataset", progress = datasetMd.checkpoint.progress, timestamp = datasetMd.checkpoint.timestamp
+
+    if err =? (await self.deleteBatch(treeCid, datasetMd)).errorOption:
+      error "Error occurred during deletion of a dataset", msg = err.msg
+    else:
+      info "Dataset deletion complete"
+  except CatchableError as err:
+    error "Unexpected error during dataset deletion", msg = err.msg, treeCid = treeCid
+
```

Comment:
So there's a quite a bit of stuff going on in `deleteBatch` and even though this method doesn't raise exception explicitly, it could happen that some code that's reachable by this method is either already raising exception or will do it in the future when the code evolves. Having that in mind, note that `superviseDatasetDeletion` is running on a newly spawned coroutine (using `asyncSpawn`) so the error is not propagated to the parent coroutine. So without this `try-catch` we are open to having an error that would never be logged if it would occur.

Reply:
> some code that's reachable by this method
Is there any code that is reachable by the wrapped try/catch that could possibly raise an exception?  I can't see anything that would raise an exception here that's not already handled. Imo, it's better to write code for what we have now, and not plan too far in the future.

> note that superviseDatasetDeletion is running on a newly spawned coroutine (using asyncSpawn) so the error is not propagated to the parent coroutine
I think this is a moot point because if there is not possibly way something in the try/catch could raise an exception, then how are we are risk of not logging an error at the callsite?

Reply:
Looking at this again, I believe there could be a `CancelledError` raised by the `await`, however I would suggest catching that error explicitly instead of `CatchableError`, as it is more illustrative of the expected behaviour to the reader. Once caught, i think it can be safely discarded, as I don't see the value in logging that information, and bubbling the `CancelledError` to the callsite will not have any effect since, as you've mentioned, this routine is being asyncSpawned.

Reply:
When we finally implement chronos v4, we can add exception tracking to async procs. And in this case, we could add `{.async:(raises:[]).}` to `deleteBatch` then the compiler will tell us for sure that it will not raise a `CancelledError`.

Reply:
This:

```nim
if err =? catch(await self.deleteBatch(treeCid, datasetMd)).errorOption:
      error "Error occurred during deletion of a dataset", msg = err.msg
```

Will discard an error returned (not thrown) by `deleteBatch`.

Reply:
Nice, I didn't know about this:

```nim
if err =? catch(await self.deleteBatch(treeCid, datasetMd)).errorOption:
      error "Error occurred during deletion of a dataset", msg = err.msg
```

Reply:
Now that we have chronos v4 support in, we can inform the compiler of what
exceptions we expect to potentially arise from this proc. For example, you could
specify that this proc only raises `async:(raises:[CancelledError])`, and then
remove the `except CatchableError` and the compiler will barf if there is a
leaked exception. Just a suggestion, but imo, very useful.